# 4.7: Deriving New Variables

## This script includes the following points:

#### 1. Import the data sets.
#### 2. Create “price_label” column.
#### 3. Create "busiest_day" column.
#### 4. Update "busiest day" column to "busiest days" to identify two "Busiest days" and two "Slowest days" of the week.
#### 5. Check the values of "busiest days" column for accuracy with observations.
#### 6. Create a new column "busiest_period_of_day" to identify the busiest hours of the day.
#### 7. Print the frequency for "busiest_period_of_day" column.
#### 8. Export the data as a pickle file.

In [3]:
# Import libraries
import pandas as pd
import numpy as np
import os

### 1. Import the data sets.

In [6]:
path = r'/Users/test/Desktop/Data Analysis/11-2024 Instacart Basket Analysis'

In [8]:
# Import ords_prods_merge dataframe from Task 4.6.2
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

### 2. Create "price_label" column.

In [11]:
# Create a subset of first million rows
df = df_ords_prods_merge[:1000000]

In [13]:
# Define a function for price_lable
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [15]:
# Apply the function.
df['price_range'] = df.apply(price_label, axis=1)

/var/folders/23/dwh_pxj57x19sw0ns77c_swh0000gn/T/ipykernel_96600/1771486740.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [16]:
# Check the values in new column.
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    652638
Low-range product    338018
High range             9344
Name: count, dtype: int64

In [17]:
# Check the price range
df['prices'].describe()

count    1000000.000000
mean           7.738643
std            4.224603
min            1.000000
25%            3.800000
50%            8.000000
75%           11.300000
max           24.500000
Name: prices, dtype: float64

In [21]:
# Apply price range using the predefined loc() Function 
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

/var/folders/23/dwh_pxj57x19sw0ns77c_swh0000gn/T/ipykernel_96600/1252634019.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [23]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [25]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [27]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     652638
Low-range product     338018
High-range product      9344
Name: count, dtype: int64

In [29]:
# Use loc() method on entire ords_prods_merge dataframe
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [30]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [33]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [35]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      412551
Name: count, dtype: int64

### 3. Create "busiest_day" column.

In [40]:
# Find frequency of orders_day_of_week.
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6203329
1    5659298
6    4495887
2    4213105
5    4205076
3    3839865
4    3783172
Name: count, dtype: int64

In [42]:
# Use for-loop to determine whether orders are on "busiest day" (0 = Saturday), "least busy" (4 = Wednesday), or "regularly busy" (other days of the week).

result = []
for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [44]:
result

['Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Bus

In [46]:
# Create a new column within ords_prods_merge dataframe and set it equal to result.
df_ords_prods_merge['busiest_day'] = result

In [48]:
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22413231
Busiest day        6203329
Least busy         3783172
Name: count, dtype: int64

### 4. Update "busiest day" column to "busiest days" to identify two "Busiest days" and two "Slowest days" of the week.

In [52]:
# Two busiest days (0, 1), two slowest days (4, 3), and regularly busy (all other days).

In [54]:
# Create column “busiest_days" from "orders_day_of_week".

result_2 = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value == 1:
    result_2.append("Busiest days")
  elif value == 4 or value == 3:
    result_2.append("Slowest days")
  else:
    result_2.append("Regularly busy")

In [56]:
result_2

['Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Slowest days',
 'Slowest days',
 'Busiest days',
 'Regularly busy',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',

### 5. Check values of new "busiest days" column for accuracy with observations.

In [61]:
# Create new column 'busiest_days' within ords_prods_merge dataframe and set it equal to result_2.
df_ords_prods_merge['busiest_days'] = result_2

In [63]:
df_ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12914068
Busiest days      11862627
Slowest days       7623037
Name: count, dtype: int64

In [67]:
# Check the output.
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  first_order  add_to_cart_order  reordered _merge  \
0                     3.0        False                  5          0   both   
1                    20.0        False                  1          1   both   
2                     6.0        False                 20          0   both   
3                     NaN         True                 10          0   both   
4                     9.0        False                 11          1   both   

     price_range_loc     busiest_day    busiest_days  
0  Mid-range product  Regularly busy  Regularly busy  
1  Mid-range product  Regularly busy  Regularly busy  
2  Mid-range product     Busiest day    Busiest days  
3  Mid-range product  Regularly busy    Slowest days  
4  Mid-range product      Least busy    Slowest days

The total value counts in the ords_prods_merge dataframe match the combined sum of all categories ("Regularly busy," "Busiest days," and "Slowest days"). For instance, "Busiest days" totals 11862627, which equals the sum of orders_day_of_week 0 and 1 (6203329 + 5659298).

### 6. Create a new column "busiest_period_of_day" to identify the busiest hours of the day.

In [71]:
# Check value counts in "order_hour_of_day" column.
df_ords_prods_merge['order_hour_of_day'].value_counts()

order_hour_of_day
10    2761333
11    2735694
14    2688728
15    2661718
13    2660570
12    2618104
16    2534744
9     2453842
17    2087273
8     1717863
18    1636226
19    1258076
20     976000
7      890923
21     795528
22     634159
23     402272
6      290450
0      218742
1      115683
5       87944
2       69360
4       53232
3       51268
Name: count, dtype: int64

The value counts above are divided into three equal groups: "Most orders" includes 10, 11, 14, 15, 13, 12, 16, and 9; "Fewest orders" includes 23, 6, 0, 1, 5, 2, 4, and 3; "Average orders" consists of all other values in order_hour_of_day.

In [74]:
# Use for-loop to label three periods of time as “Most orders,” “Average orders,” and “Fewest orders.”

result_3 = []

for value in df_ords_prods_merge["order_hour_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9]:
    result_3.append("Most orders")
  elif value in [23, 6, 0, 1, 5, 2, 4, 3]:
    result_3.append("Fewest orders")
  else:
    result_3.append("Average orders")

In [76]:
result_3

['Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Fewest orders',
 'Average orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Fewest orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most ord

In [78]:
# Create "busiest_period_of_day" column within ords_prods_merge.
df_ords_prods_merge['busiest_period_of_day'] = result_3

### 7. Print frequency of "busiest_period_of_day" column.

In [83]:
# Print value counts in "busiest_period_of_day" column.
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21114733
Average orders     9996048
Fewest orders      1288951
Name: count, dtype: int64

### 8. Export the dataframe as a pickle file.

In [88]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_derived.pkl'))